edb/pgsql/compiler/relgen.py

```diff
@@ -3859,3 +3889,219 @@ def process_encoded_param(
+                    parsed_query_pg, an_id=obj_id, ctx=inner_ctx
+                )
+                parsed_query = pathctx.get_path_var(
+                    inner_ctx.rel, parsed_query_id, aspect='value', env=ctx.env
+                )
+
+                score_pg = pgast.FuncCall(
+                    name=('pg_catalog', 'ts_rank'),
+                    args=[fts_document, parsed_query]
+                )
+                where_clause = pgast.Expr(
+                    lexpr=fts_document,
+                    name='@@',
+                    rexpr=parsed_query
+                )
+
+            pathctx.put_path_var(
+                inner_ctx.rel, out_score_id, score_pg, aspect='value'
+            )
+            inner_ctx.rel.where_clause = astutils.extend_binop(
+                inner_ctx.rel.where_clause, where_clause
+            )
+
+            in_rvar = relctx.new_rel_rvar(
+                ir_set, inner_ctx.rel, ctx=newctx
+            )
+            relctx.include_rvar(
+                newctx.rel, in_rvar, out_score_id, aspects={'value'}, ctx=newctx
+            )
+
+        obj_id_pg_ref = pathctx.get_rvar_path_var(
+            obj_rvar, out_obj_id, aspect='value', env=newctx.env
+        )
+        score_pg_ref = pathctx.get_path_var(
+            newctx.rel, out_score_id, aspect='value', env=newctx.env
+        )
+
+        tuple_expr = pgast.TupleVar(
+            elements=[
+                pgast.TupleElement(
+                    path_id=out_obj_id,
+                    name='object',
+                    val=obj_id_pg_ref,
+                ),
+                pgast.TupleElement(
+                    path_id=out_score_id,
+                    name='score',
+                    val=score_pg_ref,
+                ),
+            ],
+            named=True,
+            typeref=ir_set.typeref,
+        )
+
+        pathctx.put_path_var(
+            newctx.rel, ir_set.path_id, tuple_expr, aspect='value'
+        )
+
+        var = pathctx.maybe_get_path_var(
+            newctx.rel,
+            obj_id,
+            aspect='serialized',
+            env=newctx.env
+        )
+        if var is not None:
+            pathctx.put_path_var(
+                newctx.rel,
+                out_obj_id,
+                var,
+                aspect='serialized',
+            )
+
+    pathctx.put_path_id_map(newctx.rel, out_obj_id, obj_id)
+
+    aspects = {'value'}
+
+    func_rvar = relctx.new_rel_rvar(ir_set, newctx.rel, ctx=ctx)
+    relctx.include_rvar(
+        ctx.rel, func_rvar, ir_set.path_id, aspects=aspects, ctx=ctx
+    )
+
+    return new_stmt_set_rvar(ir_set, ctx.rel, aspects=aspects, ctx=ctx)
+
+
+def create_subrel_for_expr(
+    expr: pgast.BaseExpr,
+    *,
+    alias: str = 'expr',
+    an_id: irast.PathId,
+    ctx: context.CompilerContextLevel
+) -> irast.PathId:
+    """
+    Creates a dummy relation that contains the given expression.
+    That expr should not use any references to other relvars, since it will be
```

Comment:
I think that where this is used, the `expr` sometimes might have references to other rvars, if any nontrivial computation is done to produce the language or the query?

I don't think that should be a big problem, though? If we specify `lateral=True` when we create the rvar, it should be able to refer to the earlier things.

Reply:
When expr gets complex, this subrel gets joined in with `CROSS LATERAL JOIN`. So this is not a problem, apparently.

I'll also remove the comment saying it should not work.